packages/flutter/lib/src/widgets/page_view.dart

```diff
@@ -682,10 +699,22 @@ class _PageViewState extends State<PageView> {
         if (notification.depth == 0 && widget.onPageChanged != null && notification is ScrollUpdateNotification) {
           final PageMetrics metrics = notification.metrics;
           final int currentPage = metrics.page.round();
-          if (currentPage != _lastReportedPage) {
-            _lastReportedPage = currentPage;
+          if (currentPage != _lastOnPageChangedIndex) {
+            _lastOnPageChangedIndex = currentPage;
             widget.onPageChanged(currentPage);
           }
+        } else if (notification.depth == 0 && widget.pageDidChange != null && notification is ScrollEndNotification) {
```

Comment:
if pagesnapping is false, this will only trigger when user release the touch, that mean it will not trigger when user scroll pass the page and still holding down the finger. Based on the documentation of this callback, this might be correct because it mentioned this callback should only be called after scroll event ends. However, it does not fit the use case that you mentioned in https://github.com/flutter/flutter/issues/28096

If you want to disable scrolling after user scroll to certain page, using this call back will likely end up stuck in between pages if pagesnapping is false.

Reply:
Yes you are right, that's why I added the [WIP] label in title.
Just finished it.That's we I have done: Just close it when `pageSnapping is false`.
.Add a new property -`isBeingDragged` to `ScrollController` to indicate whether user is dragged the scrollView, and add related test case.

Now it works this way:
- If user release finger and then quickly hold the page, it won't trigger.
- If `pageSnapping` is false, never triggers.
- Only triggers when a user's finger is up and the page change movement has ended.

Reply:
I think we can ignore isBeingDragged and just check if (!widget.pageSnapping).

I might be missing something, but the only case i can think of isBeingDragged is true and currentPage != _lastPageDidChangeIndex and notification is ScrollEndNotification is when user immediately release and hold, and then time it at the right moment to drag to the perfect offset to the next page. This is an extreme corner case, and even then I think it is ok to invoke the callback. It does not worth to expose a new api in ScrollController to handle this case.

Reply:
It should still pass the test 'PageDidChange does not trigger when user is dragging' even if without isBeingDragged because ScrollEndNotification should only happened after user release his/her finger.

Reply:
Oh that’s right.So should it be kept?I think I still need add some code to it to make sure user finger’s quickly up and down works properly.

Reply:
But just image you are now in page 3, if you drag the PageView back to page 2 a little bit, release finger and hold.Then pageDidChange will be called...maybe this case is not that corner.But indeed,api should not be added too casually.Anyway I will work around it tomorrow. It ‘s really late in my time😂

Reply:
how about we listen on both ScrollEndNotification and ScrollUpdateNotification and invoke the call back whenever
abs(metrics.page - metrics.page.round()) < 10^-13? and currentPage != _lastPageDidChangeIndex, will this cause any unexpected behavior? In this case we can apply to  pageSnapping = false.

Does this call back has to be called at the end of animation?